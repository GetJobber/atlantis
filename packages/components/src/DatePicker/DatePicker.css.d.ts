declare const styles: {
  readonly "datePicker": string;
  readonly "bold": string;
};
export = styles;

