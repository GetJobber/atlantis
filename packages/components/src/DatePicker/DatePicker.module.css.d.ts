declare const styles: {
  readonly "react-datepicker__month--selecting-range": string;
  readonly "react-datepicker__year--selecting-range": string;
  readonly "datePickerWrapper": string;
  readonly "fullWidth": string;
  readonly "datePicker": string;
  readonly "inline": string;
  readonly "header": string;
  readonly "month": string;
};
export = styles;

