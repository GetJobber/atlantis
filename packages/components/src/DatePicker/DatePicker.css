@import "react-datepicker/dist/react-datepicker.module.css";

.datePickerWrapper {
  display: inline-block;
}

.fullWidth {
  display: block;
  width: 100%;
}

.datePicker:global(.react-datepicker) {
  --datepicker-cell-size: calc(var(--space-base) * 2.375);
  box-shadow: var(--shadow-base);
  border-color: var(--color-border);
  border-radius: var(--radius-base);
  font-family: var(--typography--fontFamily-normal);
}

.datePicker.inline {
  box-shadow: none;
}

:global(.react-datepicker__input-container),
:global(.react-datepicker-wrapper) {
  max-width: 100%;
}

.datePickerWrapper :global(.react-datepicker-popper) {
  z-index: var(--elevation-datepicker);
}

.datePicker :global(.react-datepicker__header) {
  padding: 0;
  border-bottom: var(--border-base) solid var(--color-border);
  background: var(--color-surface);
}

.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.datePicker :global(.react-datepicker__day) {
  display: inline-flex;
  width: var(--datepicker-cell-size);
  height: var(--datepicker-cell-size);
  margin: 0;
  border-radius: var(--radius-base);
  color: var(--color-heading);
  font-size: var(--typography--fontSize-base);
  line-height: inherit;
  background: var(--color-surface);
  align-items: center;
  justify-content: center;
}

.datePicker :global(.react-datepicker__day--outside-month) {
  border-radius: 0;
  color: var(--color-text);
  background: var(--color-surface--background);
}

.datePicker :global(.react-datepicker__month) {
  margin: 0;
}

.datePicker :global(.react-datepicker__day--selected) {
  color: var(--color-text--reverse);
  background: var(--color-interactive);
}

.datePicker :global(.react-datepicker__day-names) {
  display: flex;
  justify-content: space-between;
  margin: 0;
  padding: var(--space-smallest) 0 0;
  color: var(--color-text);
  text-transform: uppercase;
}

.datePicker :global(.react-datepicker__day-name) {
  width: auto;
  color: var(--color-text);
  flex: 1;
}

.datePicker :global(.react-datepicker__day):hover,
.datePicker :global(.react-datepicker__day):focus {
  background: var(--color-surface--hover);
}

.datePicker :global(.react-datepicker__day):focus {
  position: relative;
  box-shadow: var(--shadow-focus);
  outline: none;
}

.datePicker :global(.react-datepicker__day--selected):hover,
.datePicker :global(.react-datepicker__day--selected):focus {
  background: var(--color-interactive--hover);
}

.month {
  margin-top: var(--space-minuscule);
  color: var(--color-heading);
}
