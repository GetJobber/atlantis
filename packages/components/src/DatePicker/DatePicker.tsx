import React, { ReactElement, useState } from "react";
import ReactDatePicker from "react-datepicker";
import { XOR } from "ts-xor";
import styles from "./DatePicker.css";
import { strFormatDate } from "../FormatDate";
import { Button } from "../Button";

/**
 * Disabling no-internal-modules here because we need
 * to reach into the package to get the css file.
 */
// eslint-disable-next-line import/no-internal-modules
import "react-datepicker/dist/react-datepicker.css";

export interface DatePickerReturnedDates {
  readonly formatted: string;
  readonly raw: Date;
}

interface BaseDatePickerProps {
  /**
   * Change handler that will return the date selected.
   */
  onChange(val: DatePickerReturnedDates): void;
}

interface DatePickerModalProps extends BaseDatePickerProps {
  /**
   * Use a custom activator to trigger the DatePicker
   */
  readonly activator?: ReactElement;
}

interface DatePickerInlineProps extends BaseDatePickerProps {
  readonly inline?: boolean;
}

type DatePickerProps = XOR<DatePickerModalProps, DatePickerInlineProps>;

export function DatePicker({ onChange, activator, inline }: DatePickerProps) {
  const [startDate, setStartDate] = useState(new Date());

  return (
    <ReactDatePicker
      showPopperArrow={false}
      selected={startDate}
      className={styles.datePicker}
      inline={inline}
      onChange={handleChange}
      customInput={
        activator ? (
          activator
        ) : (
          <Button icon="calendar" ariaLabel="Open Datepicker" />
        )
      }
    />
  );

  function handleChange(val: Date) {
    setStartDate(val);
    onChange && onChange({ formatted: strFormatDate(val), raw: val });
  }
}
