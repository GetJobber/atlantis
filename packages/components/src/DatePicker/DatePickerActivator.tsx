import React, {
  ChangeEvent,
  ReactElement,
  Ref,
  cloneElement,
  forwardRef,
  isValidElement,
} from "react";
import { ReactDatePickerProps } from "react-datepicker";
import omit from "lodash/omit";
import { Button } from "../Button";

export interface DatePickerActivatorProps
  extends Pick<
    ReactDatePickerProps,
    | "value"
    | "id"
    | "ariaDescribedBy"
    | "ariaInvalid"
    | "ariaLabelledBy"
    | "ariaRequired"
  > {
  readonly activator?:
    | ReactElement
    | ((props: DatePickerActivatorProps) => ReactElement);
  readonly fullWidth?: boolean;
  onBlur?(): void;
  onChange?(
    event?: ChangeEvent<
      HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
    >,
  ): void;
  onClick?(): void;
  onFocus?(): void;
  onKeyDown?(): void;
}

export const DatePickerActivator = forwardRef(InternalActivator);

function InternalActivator(
  props: DatePickerActivatorProps,
  ref: Ref<HTMLElement>,
) {
  const { activator, fullWidth } = props;
  const newActivatorProps = omit(props, ["activator", "fullWidth"]);

  if (activator) {
    if (isValidElement(activator)) {
      const isAComponent = typeof activator.type === "function";

      return cloneElement(activator, {
        ...newActivatorProps,
        ...(isAComponent && { fullWidth: fullWidth }),
        // @ts-expect-error - Issue with react types not including `ref` in
        // cloneElement. https://github.com/DefinitelyTyped/DefinitelyTyped/issues/40888
        ref,
      });
    } else if (typeof activator === "function") {
      return activator(props);
    } else {
      return null;
    }
  } else {
    return (
      <Button
        variation="work"
        type="tertiary"
        icon="calendar"
        ariaLabel="Open Datepicker"
        fullWidth={fullWidth}
        {...newActivatorProps}
      />
    );
  }
}
