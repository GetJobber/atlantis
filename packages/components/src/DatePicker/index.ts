export { DatePicker } from "./DatePicker";

export * from "./RawDatePicker";
