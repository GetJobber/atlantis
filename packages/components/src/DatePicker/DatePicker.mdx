---
name: DatePicker
menu: Components
route: /components/date-picker
showDirectoryLink: true
---

import { useState } from "react";
import { Playground, Props } from "docz";
import { DataDump } from "@jobber/components/DataDump";
import { Content } from "@jobber/components/Content";
import { Button } from "@jobber/components/Button";
import { DatePicker } from ".";

# Date Picker

A `DatePicker` is used to allow a user to select a date from a visual calendar
interface.

While this component can be used in isolation, the intention is that in many
cases it will supplement an input component to provide the user with a
multi-mode opportunity to select a date.

```ts
import { DatePicker } from "@jobber/components/DatePicker";
```

<Playground>
  {() => {
    const [date, setDate] = useState();
    return (
      <Content>
        <DatePicker selected={date} onChange={setDate} />
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Playground>

## Props

<Props of={DatePicker} />

## Design & Usage Guidelines

Use Datepicker when you are looking to provide the user with a highly visual
interface for intuitive date selections.

A classic use case for a standalone Datepicker would be as a means to navigate
within a larger calendar interface.

It may be beneficial to supplement the Datepicker with alternate means of
entering their date selection, such as providing a `date` input.

### Inline

Use an inline Datepicker when you have sufficient space in the interface to have
a permanent date selection fixture, or when date selection is the primary
function of the interface, such as a stepped flow where the user is being asked
a series of scheduling questions individually, in sequence.

In cases like complex scheduling forms where all the factors of a schedule such
as times, repeating frequency, and duration are being asked, consider using the
default popover-type behaviour for those use cases to reduce an overwhelming
volume of UI elements to process at once.

<Playground>
  {() => {
    const [date, setDate] = useState();
    return (
      <Content>
        <div>
          <DatePicker selected={date} inline onChange={setDate} />
        </div>
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Playground>

## Custom Activator

If you want to use a different activator to display the `DatePicker` you can
pass a `React` component as the `activator`.

<Playground>
  {() => {
    const [date, setDate] = useState();
    return (
      <Content>
        <DatePicker
          selected={date}
          onChange={setDate}
          activator={<Button label="My fancy activator" variation="learning" />}
        />
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Playground>

## Disabling dates

If you want to enable for selection only for a set of dates you can pass
`minDate` or `maxDate` to the component as props.

<Playground>
  {() => {
    const [date, setDate] = useState(new Date("02/17/2022"));
    return (
      <Content>
        <div>
          <DatePicker
            minDate={new Date("02/07/2022")}
            maxDate={new Date("02/27/2022")}
            selected={date}
            inline
            onChange={setDate}
          />
        </div>
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Playground>

## Accessibility

At this point, the `DatePicker` component is a wrapped version of
[React DatePicker](https://reactdatepicker.com/). This means that the
accessibility of the component is handled by the `React DatePicker` component.

Users can operate all of the controls by keyboard, including the `tab` to cycle
between Next/Previous month and the calendar, arrow keys to navigate between
dates, and `space` and/or `return` to make selections. Pressing `escape` will
close the Datepicker.

The date available for selection is read aloud to assistive technology in the
format "Choose {date} (button)". For example, "Choose December 1st, 2021
(button)".

If using a custom activator, ensure that the activator is keyboard-operable.

## Related Components

- If you are looking to use the `DatePicker` in a `Form`, consider the
  [InputDate](/components/input-date) component.
- For a time input, use [InputTime](/components/input-time)
- To present dates that have already been selected in a consistent format, use
  [FormatDate](/components/date-formatter)
- To represent past or future dates and times, use
  [FormatRelativeDateTime](/components/format-relative-date-time)

## Notes

As mentioned above, the Datepicker wraps
[React DatePicker](https://reactdatepicker.com/).
