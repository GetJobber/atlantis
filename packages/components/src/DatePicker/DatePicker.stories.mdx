import { useState } from "react";
import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { DataDump } from "@jobber/components/DataDump";
import { Content } from "@jobber/components/Content";
import { Button } from "@jobber/components/Button";
import { DatePicker } from "@jobber/components/DatePicker";

<Meta title="Components/Selections/DatePicker" component={DatePicker} />

# Date Picker

A `DatePicker` is used to allow a user to select a date from a visual calendar
interface.

While this component can be used in isolation, the intention is that in many
cases it will supplement an input component to provide the user with a
multi-mode opportunity to select a date.

```ts
import { DatePicker } from "@jobber/components/DatePicker";
```

<Canvas>
  <Story name="DatePicker">
    {args => {
      const [date, setDate] = useState(args.date);
      return (
        <Content>
          <DatePicker {...args} selected={date} onChange={setDate} />
          {date && <DataDump defaultOpen data={date} />}
        </Content>
      );
    }}
  </Story>
</Canvas>

<ArgsTable of={DatePicker} story="DatePicker" />

## Design & usage guidelines

Use Datepicker when you are looking to provide the user with a highly visual
interface for intuitive date selections.

A classic use case for a standalone Datepicker would be as a means to navigate
within a larger calendar interface.

It may be beneficial to supplement the Datepicker with alternate means of
entering their date selection, such as providing a `date` input.

### Inline

Use an inline Datepicker when you have sufficient space in the interface to have
a permanent date selection fixture, or when date selection is the primary
function of the interface, such as a stepped flow where the user is asked a
series of scheduling questions individually, in sequence.

In cases like complex scheduling forms where all the factors of a schedule such
as times, repeating frequency, and duration are being asked, consider using the
default popover-type behaviour for those use cases to reduce an overwhelming
volume of UI elements to process at once.

<Canvas>
  {() => {
    const [date, setDate] = useState();
    return (
      <Content>
        <div>
          <DatePicker selected={date} inline onChange={setDate} />
        </div>
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Canvas>

## Custom activator

If you want to use a different activator to display the `DatePicker` you can
pass a `React` component as the `activator`.

<Canvas>
  {() => {
    const [date, setDate] = useState();
    return (
      <Content>
        <DatePicker
          selected={date}
          onChange={setDate}
          activator={<Button label="My fancy activator" variation="learning" />}
        />
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Canvas>

## Restricting date ranges

If you want to enable for selection only a range of dates you can pass `minDate`
or `maxDate` to the component as props.

<Canvas>
  {() => {
    const [date, setDate] = useState(new Date("02/17/2022"));
    return (
      <Content>
        <div>
          <DatePicker
            minDate={new Date("02/07/2022")}
            maxDate={new Date("02/27/2022")}
            selected={date}
            inline
            onChange={setDate}
          />
        </div>
        {date && <DataDump defaultOpen data={date} />}
      </Content>
    );
  }}
</Canvas>

## Highlighted dates

If you need to call attention to a series of dates on the calendar, the
`highlightDates` prop accepts an array of dates that will be displayed with
additional emphasis.

<Canvas>
  {() => {
    const [date, setDate] = useState(new Date("01/02/2023"));
    const highlightDates = [
      new Date("01/01/2023"),
      new Date("02/20/2023"),
      new Date("04/07/2023"),
      new Date("05/22/2023"),
      new Date("07/01/2023"),
      new Date("08/07/2023"),
      new Date("09/04/2023"),
      new Date("10/09/2023"),
      new Date("11/13/2023"),
      new Date("12/25/2023"),
      new Date("12/26/2023"),
    ];
    return (
      <Content>
        <DatePicker
          minDate={new Date("01/01/2023")}
          maxDate={new Date("12/31/2023")}
          selected={date}
          inline
          onChange={setDate}
          highlightDates={highlightDates}
        />
        <DataDump defaultOpen label="highlightDates" data={highlightDates} />
      </Content>
    );
  }}
</Canvas>

## Accessibility

At this point, the `DatePicker` component is a wrapped version of
[React DatePicker](https://reactdatepicker.com/). This means that the
accessibility of the component is handled by the `React DatePicker` component.

Users can operate all of the controls by keyboard, including the `tab` to cycle
between Next/Previous month and the calendar, arrow keys to navigate between
dates, and `space` and/or `return` to make selections. Pressing `escape` will
close the Datepicker.

The date available for selection is read aloud to assistive technology in the
format "Choose {date} (button)". For example, "Choose December 1st, 2021
(button)".

If using a custom activator, ensure that the activator is keyboard-operable.

## Related components

- If you are looking to use the `DatePicker` in a `Form`, consider the
  [InputDate](../?path=/docs/components-forms-and-inputs-inputdate--input-date)
  component.
- For a time input, use
  [InputTime](../?path=/docs/components-forms-and-inputs-inputtime--input-time)
- To present dates that have already been selected in a consistent format, use
  [FormatDate](../?path=/docs/components-utilities-formatdate--format-date)
- To represent past or future dates and times, use
  [FormatRelativeDateTime](../?path=/docs/components-utilities-formatrelativedatetime--format-relative-date-time)

## Notes

As mentioned above, the Datepicker wraps
[React DatePicker](https://reactdatepicker.com/).
