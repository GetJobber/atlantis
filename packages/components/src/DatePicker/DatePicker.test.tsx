import React from "react";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { DatePicker } from "./DatePicker";

afterEach(cleanup);

beforeEach(() => {
  /**
   * We need to mock our date because DatePicker sets an initial date
   * to `new Date`, we could only predict the current date for a month.
   *
   * This will mock our system time to `April 7, 2021`.
   */
  jest.useFakeTimers("modern");
  jest.setSystemTime(new Date(2021, 3, 7));
});

it("renders only a button by default", () => {
  const { container, queryByText } = render(
    <DatePicker onChange={jest.fn()} />,
  );

  /**
   * Checking for null here, as if someone was to change
   * the `inline` prop to true by default this would fail.
   */
  expect(queryByText("15")).toBeNull();
  expect(container).toMatchSnapshot();
});

it("returns the dates from onChange", async () => {
  const changeHandler = jest.fn();
  const { getByTestId, getByText } = render(
    <DatePicker onChange={changeHandler} />,
  );

  const button = getByTestId("calendar");
  fireEvent.click(button);

  const date = getByText("15");
  fireEvent.click(date);

  /**
   * This gets the first result from the `changeHandler` mock
   * now that it has been called
   */
  const [[changeHandlerResult]] = changeHandler.mock.calls;

  expect(changeHandlerResult).toMatchObject({
    raw: expect.any(Date),
    formatted: "Apr 15, 2021",
  });
});

it("allows for a custom activator to open the DatePicker", () => {
  const { getByText } = render(
    <DatePicker onChange={jest.fn()} activator={<div>Activate me</div>} />,
  );

  const button = getByText("Activate me");
  fireEvent.click(button);

  expect(getByText("15")).toBeInstanceOf(HTMLDivElement);
});

it("always appears when inline", () => {
  const { getByText } = render(<DatePicker onChange={jest.fn()} inline />);

  expect(getByText("15")).toBeInstanceOf(HTMLDivElement);
});
