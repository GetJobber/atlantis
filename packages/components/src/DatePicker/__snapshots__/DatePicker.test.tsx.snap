// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders only a button by default 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <button
        aria-label="Open Datepicker"
        class="button base onlyIcon work primary"
        type="button"
      >
        <svg
          class="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
          data-testid="calendar"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M640 170.667c0-23.564 19.102-42.667 42.667-42.667s42.667 19.103 42.667 42.667v42.667h85.333c47.13 0 85.333 38.205 85.333 85.333v512c0 47.13-38.204 85.333-85.333 85.333h-597.333c-47.128 0-85.333-38.204-85.333-85.333v-512c0-47.128 38.205-85.333 85.333-85.333h85.333v-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667v42.667h256v-42.667zM213.333 298.667v85.333h597.333v-85.333h-597.333zM810.667 469.333h-597.333v341.333h597.333v-341.333z"
          />
        </svg>
        <span
          class="base extraBold small uppercase"
        />
      </button>
    </div>
  </div>
</div>
`;
